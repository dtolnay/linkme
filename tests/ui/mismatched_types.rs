#![cfg_attr(feature = "used_linker", feature(used_with_arg))]

use linkme::{disjointed_static, distributed_slice};

pub struct Bencher;
pub struct Config;

#[distributed_slice]
pub static BENCHMARKS: [fn(&mut Bencher)];

#[disjointed_static]
pub static DEFAULT_CONFIG: Config;

#[distributed_slice(BENCHMARKS)]
static BENCH_WTF: usize = 999;

#[disjointed_static(DEFAULT_CONFIG)]
static CONFIG_WTF: usize = 999;

#[distributed_slice(BENCHMARKS)]
fn wrong_bench_fn<'a>(_: &'a mut ()) {}

fn main() {}
