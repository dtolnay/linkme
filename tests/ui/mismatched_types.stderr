error[E0308]: mismatched types
  --> tests/ui/mismatched_types.rs:15:19
   |
14 | #[distributed_slice(BENCHMARKS)]
   | -------------------------------- arguments to this function are incorrect
15 | static BENCH_WTF: usize = 999;
   |                   ^^^^^ expected fn pointer, found `usize`
   |
   = note: expected fn pointer `fn() -> &'static for<'a> fn(&'a mut Bencher)`
              found fn pointer `fn() -> &'static usize`
note: method defined here
  --> src/distributed_slice.rs
   |
   |     pub unsafe fn private_typecheck(self, get: fn() -> &'static T) {
   |                   ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/mismatched_types.rs:18:20
   |
17 | #[disjointed_static(DEFAULT_CONFIG)]
   | ------------------------------------ arguments to this function are incorrect
18 | static CONFIG_WTF: usize = 999;
   |                    ^^^^^ expected `Config`, found `usize`
   |
   = note: expected fn pointer `fn() -> &'static Config`
              found fn pointer `fn() -> &'static usize`
note: method defined here
  --> src/disjointed_static.rs
   |
   |     pub const fn private_typecheck(self, get: fn() -> &'static T) {
   |                  ^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/mismatched_types.rs:21:1
   |
20 | #[distributed_slice(BENCHMARKS)]
   | -------------------------------- arguments to this function are incorrect
21 | fn wrong_bench_fn<'a>(_: &'a mut ()) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Bencher`, found `()`
   |
   = note: expected fn pointer `fn() -> &'static for<'a> fn(&'a mut Bencher)`
              found fn pointer `fn() -> &'static for<'a> fn(&'a mut ())`
note: method defined here
  --> src/distributed_slice.rs
   |
   |     pub unsafe fn private_typecheck(self, get: fn() -> &'static T) {
   |                   ^^^^^^^^^^^^^^^^^
